### DISCUSSION TITLE
Dynamic Forms

### DISCUSSION THREAD

**Discussion by fenilli** [statelyai/xstate#2964] [Q&A] [answered]
I'm new to XState, reading a lot about it, and I have the concept of creating a form machine that could have many fields and other forms, because it makes a lot of sense when thinking about state machines ( having, loading, error, and so on states ).
But when thinking on how would I make it dynamic, meaning I could add a form to a form, or add a input field to it, I can't seen to find much about in the documentation on how that is handled, if actors are better, or adding everything in an object before creating a machine.

so I question would it be better to have in the context an array, that could have values or other forms ( nested forms ), or is there a "better" solution for this problem with actors, services and so on?

**Comment by davidkpiano** [comment, marked as answer]
This is a really good question, and something we're thinking about for a future `@xstate/form`. I personally think that each field can operate as its own actor, and this is apparent from popular form libraries that allow "subscriptions" to each of the fields, which the actors in XState naturally allow.

**Reply by fenilli** [reply]
That is really cool, `@xstate/form` would be amazing to have, i will be basically doing the same for now until there is a official library.

So if each field is its own actor, he should know that onChange or onInput events need to bubble up to its own parent, by sending an event to its parent ID, and as each field is its own thing, they're actually parallel to each other.

Also a form could also spawn another nested form as an actor, and that would be theoretically unlimited, so nested forms would also bubble up their children to its parent, so everyone knows their state without actually knowing their implementation.

If I understand correct in the docs, It would make sense to use the createModel + Actors to make those in a more configurable part, because inputs could have different actions, context and even their own machine. does that make sense?

**Comment by fenilli** [comment]
After reading about a lot, I think actors wouldn't work for what I'm trying to do, basically a form builder, because spanwed actors can't be serialized, meaning that I couldn't have persistent inputs that where added to the form.
